torch/distributed/distributed_c10d.py

```diff
@@ -2326,11 +2326,6 @@ def scatter(tensor, scatter_list=None, src=0, group=None, async_op=False):
         input_tensors = [scatter_list]
         output_tensors = [tensor]
     else:
-        if scatter_list:
-            raise ValueError(
-                "Argument ``scatter_list`` must NOT be specified "
-                "on non-source ranks."
-            )
```

Comment:
Is this an unintended change?

Reply:
This is intended change actually, it's to match the behavior in `gather` where `dist.gather` allow user to pass in gather_list on non-dst ranks, but we internally treat it as empty list https://github.com/pytorch/pytorch/blob/defde3bb04e2c98765dbee2d214b024de20626d0/torch/distributed/distributed_c10d.py#L2253

I think the `dist.gather` behavior make more sense, user can call the collective on all ranks SPMD, and inside `dist.gather` we ignore the gather/scatter_list on non-dst/src ranks. Otherwise, user have to write conditional code everytime to perform gather/scatter, i.e. for scatter it must be:
```
if current_rank == src_rank:
    dist.scatter(local_tensor, scatter_list=tensors_to_scatter, src=src_rank, group=pg)
else:
    dist.scatter(local_tensor, scatter_list=None, src=src_rank, group=pg)    
```
which might be very in-convenient for user. Our low level primitive: `pg.gather/scatter` enforce the non-dst/src ranks must pass None.

Reply:
Is it possible to make this logic (making scatter/gather list to be None for non_src_rank) centralized somewhere in Python? Since we have enforced this in CPP.

Reply:
it was embedded in `dist.gather` https://github.com/pytorch/pytorch/blob/c1a4714bf1195faa1913541b190e8019f9f5dac5/torch/distributed/distributed_c10d.py#L2261 and `dist.scatter`, https://github.com/pytorch/pytorch/blob/c1a4714bf1195faa1913541b190e8019f9f5dac5/torch/distributed/distributed_c10d.py#L2334, Im just remove the error raising, are you suggesting we have a util function to deal with that? I guess that's a small piece of code only for gather/scatter, maybe we don't need to generalize it out?

Reply:
The thing is that our docs clearly say it should not be specified on src_rank:

```
        scatter_list (list[Tensor]): List of tensors to scatter (default is
            None, must be specified on the source rank)
```

So this change would be confusing for users where the docs say something whereas the actual behavior is different.

Reply:
While both `scatter` and `gather` doc does not mention that `scatter_list` "must ONLY be specified on the source rank", so I guess this part is a bit confusing to the user. Anyways, I think it should be a separate discussion around consistency of our c10d APIs, not related to this PR specifically. I just updated the PR to make sure we are passing None on non-dst/src rank, and remove this unintentional change.

Reply:
created a issue to track this conversation https://github.com/pytorch/pytorch/issues/74323